### PR TITLE
AVRO-2652: Use Tox to Enable Multi Python Testing

### PR DESCRIPTION
**Pull request by kojiromike**
### Jira

- [x] My PR addresses [AVRO-2652](https://issues.apache.org/jira/browse/AVRO-2652)
- [x] ...and references it in the PR title.
- [x] Adds `tox` as a test meta-runner. [Tox has an MIT license](https://github.com/tox-dev/tox/blob/f1a933b59c2682f40054c0cabe8c1f42a3635168/LICENSE). That license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Tests

- [x] My PR changes how we run tests, but doesn't really change the tests.

### Commits

- [x] My commits all reference Jira issues in their subject lines.
- [x] In addition, my commits follow the guidelines from "[How to write a good git commit message](https://chris.beams.io/posts/git-commit/)".

### Documentation

- [x] My PR contains no documentation changes because it does not change how the avro implementation functions.

### CODE REVIEW

**Comment by Fokko** [comment]
I never understood why we need tox when we have Docker. I understand that we now have one single environment for everything which isn't ideal.

**Comment by kojiromike** [comment]
By itself tox doesn't do much, and if we had a bunch of different Dockerfiles it might be a different story, but in this case, with python 2.7 and 3.5 installed in the same image, tox will be pretty useful. I am laying a groundwork here to be able to run multiple python versions against the same lang/py tests to show that we have effective python 3.5 support and can deprecate lang/py3.

**Comment by kojiromike** [comment]
Another useful thing tox does is actually run the tests against the installable package, instead of testing against just a bunch of files. I found several bugs in our packaging setup just by trying to enable tox.

**Comment by RyanSkraba** [comment]
The additional fix for AVRO-2657 works for me (in the build logs:)
```
+ cd lang/py
+ ./build.sh interop-data-test
TEST INTEROP
============
READING perl_deflate.avro
READING perl_zstandard.avro
READING perl.avro
READING csharp_deflate.avro
<snip>
READING py_zstandard.avro
READING py_snappy.avro
READING py_deflate.avro
...
```

File: lang/py/tox.ini [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    python-snappy
+    zstandard
+whitelist_externals =
+    cp
+    mkdir
+commands_pre =
+    mkdir -p avro/test/interop {toxinidir}/../../build/interop/data
+    cp -r {toxinidir}/../../build/interop/data avro/test/interop
+    python -m avro.test.gen_interop_data avro/interop.avsc avro/test/interop/data/py.avro
+    cp -r avro/test/interop/data {toxinidir}/../../build/interop
```
Comment (kojiromike) [inline lang/py/tox.ini]: @sekikn @RyanSkraba I've rolled the AVRO-2657 fix into this PR, using slightly different implementations in build.sh and tox.ini. You can see here the precommand always does a best-effort to get any avro generated by other languages' interop generators into the python data path. Unlike `./build.sh interop-data-test`, this approach will test with any environment that we can describe to tox (see #744). Here, the avro files don't get copied into the working copy, but only into the tox virtualenv.

Do you have any feedback on this approach?

**Comment by sekikn** [comment]
@kojiromike Sorry for the late reply, and [failing CI after merging this PR](https://travis-ci.org/apache/avro/builds/627013663) is my bad. Merging #705 is supposed to resolve it.

**Comment by kojiromike** [comment]
That's ok. I merged the other pr. Looking forward to being green again!

**Comment by sekikn** [comment]
I saw the CI went back to green. Thanks @kojiromike!

**Comment by Fokko** [comment]
Sorry for the late reply. On the second thought, I think having tox in this case makes sense. I don't see any other pragmatic options.

**Pull request merged**